Brownout2

eggaskin opened the pull request:
If this works for everyone, we will use this as a base for our subsystem power management/brownout. Please review.

dx44383 committed:
Create README.md (#11)

eggaskin committed:
Upgrade to 2022 (#15)

* Create README.md (#11) (#14)

Co-authored-by: dx44383 <92555011+dx44383@users.noreply.github.com>

* Upgrade to WPILib 2022.1.1. (#13)

Co-authored-by: Ethan <e3lcool@gmail.com>
Co-authored-by: dx44383 <92555011+dx44383@users.noreply.github.com>

eggaskin committed:
current limit implementation base logic

eggaskin committed:
dynamic scaling + more structure things

proan commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,168 @@
+
+    private ControllableSubsystem tankSubsystem;
+
+//    private ControllableSubsystem climbSubsystem;
```
Alignment

Also, is there a reason you have an empty line between each subsystem? I would think they'd make a nice block of 4 or so line.

proan commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,168 @@
+
+   // private ControllableSubsystem shooterSubsystem;
+
+    private int currentCurrentLimit;
```
This is a personal style recommendation: when I name variables that correspond to physical things I like to append the units to the variable name. In this case it would be `currentCurrentLimit_A`. This way I can check units by looking at variable names.

proan commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,168 @@
+
+        this.tankSubsystem = tankSubsystem;
+
+        currentCurrentLimit = 350;
```
1. Could these be constants, or will they change during a match?
2. Are you sure you want to do integer arithmetic with these?

proan commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,168 @@
+
+        System.out.println("Checking for a brownout...");
+        // Check PDP voltage; if close to a brownout:
+        if (PDP.getVoltage() < 7) {
```
Does `PDP.getVoltage()` return an integer or a floating point number? You should make your comparison value match.

proan commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,168 @@
+        int channels = 0;
+
+        for (int channel : PDPChannel) {
+            sum += PDP.getCurrent(channel);
```
Does `PDP.getCurrent()` return an integer or a floating point number?

proan commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,168 @@
+        //scale all subsystems by set (sensible) amount
+        for (ControllableSubsystem subsystem : subsystems) {
+            int currDrawn = subsystem.getTotalCurrentDrawn();
+            int tempscale = currDrawn - (int)(currDrawn*0.2);
```
I'm concerned about all the integer arithmetic. Do we know what the expected range of values are here?

ky28059 replied:
It's hard, because `PDP.getCurrent()` returns `double`, but `CANSparkMax` expects an `int` for `setSmartCurrentLimit()`. I can get rid of some of the intermediary `int` casts but it will still have to be cast to `int` in the end for scaling.

proan replied:
Yes, let's do that. Since you measure in `double` let's use that for all the math and then cast (or better yet, design a rounding function) to an int for the current limit. CS math isn't quite the same as pencil-and-paper math, so it's best to stick with `double` for the math (that only has issues with equals, not multiplication and division).

proan commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -44,9 +45,16 @@
+  private int currentLimit;
+  private int minCurrent;
+
   public static final double ENCODER_ROTATIONS_TO_METERS = 5 / 92.08;
```
Can you also make this `5.0` please?

proan commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -212,6 +220,17 @@ public DifferentialDriveWheelSpeeds getWheelSpeeds() {
+  public void setCurrentLimit(int limit) {
+    this.currentLimit = limit;
+
+    int motorLimit = currentLimit/4; //number of motors
```
Be careful with integer division. Do you know what happens for `currentLimit` not being evenly divisible by 4?

proan reviewed (dismissed):
I'd like to see the current limits be constants at a minimum.

ky28059 committed:
Code cleanup

ky28059 commented:
Instead of having a `checkPriority(ControllableSubsystem)` method which returns an int from 1-n and _then_ having to iterate to find the lowest priority, can we instead have a `Queue<ControllableSubsystem>` in order from low to high priority and have `scale()` keep popping from that? The only reason I can think of for this maybe being a little complex is converting the array in the constructor to a Queue; maybe we can assume the subsystems are passed in in order of priority?

ky28059 commented:
Although I suppose returning an `int` priority is better for dynamically updating priority based on conditions like whether the turret is ready to shoot or whether its climb time; in that case though, I think maybe a `double` priority would allow for better tuning?

ky28059 commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,123 @@
+    for (ControllableSubsystem subsystem : subsystems) {
+      double currDrawn = subsystem.getTotalCurrentDrawn();
+      int tempscale = (int) (currDrawn * 0.8);
+      subsystem.setCurrentLimit(tempscale);
```
Why does brownout scaling set everything to a flat `0.8` scale while `scale()` sets everything to its minimums? Are there cases where the minimum value is above `0.8 * currDrawn` and the brownout scaling is less harsh than normal scaling?

proan replied:
Maybe that 80% should be a default value?
BTW this is a good design question to answer.

ky28059 commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,123 @@
+    int lowPriority = 10; // This int should be higher than the number of subsystems
+
+    // If we've already scaled many subsystems, just scale more dramatically for everything
+    if (checked.size() > 3) {
```
Why does `scale()` give up after 3 subsystems? Also, if brownout scaling is sometimes more forgiving than `minCurrent`, would this result in the lowest priority subsystems getting _more_ power than if they had been scaled normally?

ky28059 commented on src/main/java/frc/robot/PowerController.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    double currDrawn = lowestPriority.getTotalCurrentDrawn();
+    lowestPriority.setCurrentLimit((int) lowestPriority.minCurrent());
```
Seems pretty harsh, perhaps doing some math and scaling things proportionally would be better?

ky28059 commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -241,4 +247,25 @@ public void resetPosition() {
+
+  @Override
+  public double getTotalCurrentDrawn() {
+    return PowerController.getCurrentDrawnFromPDP(fLeftMotorPort, fRightMotorPort, bLeftMotorPort, bRightMotorPort);
```
Perhaps this can instead be a `getPDPPorts` method and have `PowerController` call `getCurrentDrawn` itself.

proan commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -45,16 +46,13 @@
+  private final double minCurrent = 50.0;
 
-  public static final double ENCODER_ROTATIONS_TO_METERS = 5 / 92.08;
+  public static final double ENCODER_ROTATIONS_TO_METERS = 5.0 / 92.08;
```
👍

proan commented:
Regarding priorities for your mechs/modules, @ky28059 has raised some good questions:
- Do priorities shift during a match (My thought is "yes" when you transition to "climb" but because that's the end phase, it doesn't significantly impact the priority system design)
- How to represent priority? Is it a number, a position in a list or queue, or some other kind of custom class or datatype? (An example from Linux is the udev system, which uses integers from 00 to 99 with 99 being last and thus most important)

ky28059 committed:
Change PDP to PDH :)

eggaskin committed:
more logic and prioritization

Ayeeti committed:
github is making me commit this to switch branches

Ayeeti committed:
added checkDynamicPriority and getDynamic Priority

ky28059 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -13,14 +13,21 @@
 import edu.wpi.first.wpilibj.Filesystem;
 import edu.wpi.first.wpilibj.GenericHID;
 import edu.wpi.first.wpilibj.XboxController;
+<<<<<<< HEAD
```
Github made you commit merge conflicts? Have you resolved these or are they just on the branch?

ky28059 replied:
By the way, you should keep `HEAD` for all of these. The conflicts look to be from very old changes from before pathfollowing was merged.

eggaskin committed:
1-10 priority system with base low current scaling code

eggaskin committed:
syntax and PDH

ky28059 committed:
Merge branch 'brownout2' of https://github.com/grt192/GRTCommandBased into brownout2

ky28059 committed:
Code recleanup

I think Aditi bulldozed over my old commits when merging her branch with remote, so this merge readds those commits and makes it compatible with the new changes.

ky28059 committed:
Limit based brownout scaling

ky28059 committed:
Use an abstract class for `GRTSubsystem`

This allows easier shared logic between subsystems (an interface would require each subsystem to implement their own versions of every method as it does not have access to fields). The original idea was to use this for climb cleanup as well as brownout, but basically anything that needs to be shared across subsystems can exist here.

ky28059 committed:
Fix `changePriority` logic

ky28059 committed:
Implement weighted average scaling

ky28059 commented:
The idea for weighted average (and priority setting) is for subsystems to be able to somewhat scale themselves based on current situations. For example, in `InternalSubsystem.java`, 
```java
@Override
public void periodic() {
    if (shotRequested) {
        powerController.setPriority(this, 8.0);
        powerController.setPriority(turretSubsystem, 8.0);
    } else {
        powerController.setPriority(this, 5.0);
        powerController.setPriority(turretSubsystem, 5.0);
    }
    // ...
}
```
With weights, it was detemined that `base * dynamic` would give too harsh a difference between a subsystem with base priority `4.0` and a subsystem with base priority `2.0` (here, tank would get double the allotted current as turret for just existing). However, some form of multiplication would be really useful for being able to nuke the priority of a subsystem during something like climb; `climbInit()` might set a subsystem's dynamic priority to 0 so that during climb, current isn't wasted by being allotted to a deactivated subsystem.

Or we could do something else for climb, idk.

ky28059 committed:
Use multiplication, adjust base priorities

ky28059 commented:
![image](https://user-images.githubusercontent.com/60120929/152090580-5b9d2d9a-7989-477a-840c-b4350317b116.png)
![image](https://user-images.githubusercontent.com/60120929/152090607-8c6c2527-dcc1-4c04-a5c7-828152b9413b.png)

The precise values of the base priorities can be played around with, but these values seem reasonable (comparing tank to turret when both are at resting dynamic priority and when turret is at a higher priority).

ky28059 commented:
**Alternative brownout strategy from Liang:**
Make sure at all times that the sum of the current limits of all the subsystems adds up to the sustainable current. When a subsystem begins to approach its limit (the ratio of used to unused current increases), allocate it more limit by taking limit from a subsystem whose ratio is lower. 

This would naturally fix the climb (and climbInit) problem as climb will not use any current before the climb phase which means its limit would be distributed to other users (a similar process holds true for subsystems during climb).

I can attempt to implement this soon.

ky28059 commented:
Although, an issue with this approach is that in the scenario where the current draw is pushing the sustainable current there isn't really a clean way for a subsystem to say "my ratio might be the same as this other system but screw them and give me more power"

ky28059 committed:
Implement new brownout plan, thread based updating

ky28059 committed:
Add minimum current logic

This makes the calculations a little less clean, but is needed to prevent subsystems "locking" when drawing 0 current (resulting in their limits being set to 0) as well as give each subsystem some breathing room.

ky28059 reviewed (dismissed):
I'm self-reviewing here, but I think brownout is mergeable. The new plan is really really clean!! (and its one issue with minimum and 0 currents was just fixed). I know @proan brought up the difference between sustainable current draw vs sustainable voltage, but given that we'll be using new batteries I think we can measure the sustainable current value and trust that maintaining that will sustain the voltage at a desired level for the entire match.

ky28059 committed:
Recursive `minCurrent` scaling

ky28059 committed:
Small cleanup

ky28059 committed:
Remove thread-based updating in favor of periodic

ky28059 committed:
Fix testing bugs

ky28059 reviewed (approved):
The bugs that came up during testing were fixed so hopefully everything is working swell. The `totalSustainableCurrent` needs to be measured, as does limits for each subsystem, but this branch can be merged so that I can rebase `shooter`.
